lib/rex/proto/amqp/version_0_9_1/client.rb

```diff
@@ -0,0 +1,353 @@
+          { name: 'authentication_failure_close', data: { data_type: 't'.ord, data: true } },
+          { name: 'basic.nack', data: { data_type: 't'.ord, data: true } },
+          { name: 'connection.blocked', data: { data_type: 't'.ord, data: true } },
+          { name: 'consumer_cancel_notify', data: { data_type: 't'.ord, data: true } },
+          { name: 'publisher_confirms', data: { data_type: 't'.ord, data: true } }
+        ] } }
+      ],
+      # https://www.rabbitmq.com/access-control.html#mechanisms
+      mechanism: 'PLAIN',
+      response: build_sasl_response_plain(username, password),
+      locale: locale
+    })
+  end
+
+  def recv_connection_close_ok
+    resp = recv_frame
+    unless is_method_frame?(resp, Version091::Frames::MethodArguments::AmqpVersion091ConnectionCloseOk)
+      raise Error::UnexpectedReplyError.new(resp)
+    end
+
+    resp
+  end
+
+  def recv_connection_open_ok
+    resp = recv_frame
+    unless is_method_frame?(resp, Version091::Frames::MethodArguments::AmqpVersion091ConnectionOpenOk)
+      raise Error::UnexpectedReplyError.new(resp)
+    end
+
+    resp
+  end
+
+  def recv_connection_start
+    resp = recv_frame
+    unless is_method_frame?(resp, Version091::Frames::MethodArguments::AmqpVersion091ConnectionStart)
+      raise Error::UnexpectedReplyError.new(resp)
+    end
+
+    @server_info = {
+      properties: resp.arguments.server_properties.coerce,
+      security_mechanisms: resp.arguments.mechanisms.split(' '),
+      locales: resp.arguments.locales.split(' ')
+    }
+
+    resp
+  end
+
+  def send_connection_close(arguments={})
+    conn_close = Version091::Frames::AmqpVersion091MethodFrame.new
+    conn_close.arguments = Version091::Frames::MethodArguments::AmqpVersion091ConnectionClose.new(arguments)
+    send_frame(conn_close)
+
+    nil
+  end
+
+  def send_connection_open(arguments={})
+    connection_open = Version091::Frames::AmqpVersion091MethodFrame.new
+    connection_open.arguments = Version091::Frames::MethodArguments::AmqpVersion091ConnectionOpen.new(arguments)
+    send_frame(connection_open)
+
+    nil
+  end
+
+  def send_connection_start_ok(arguments={})
+    connection_start_ok = Version091::Frames::AmqpVersion091MethodFrame.new
+    connection_start_ok.arguments = Version091::Frames::MethodArguments::AmqpVersion091ConnectionStartOk.new(arguments)
+    send_frame(connection_start_ok)
+
+    nil
+  end
+
+  def send_protocol_header
+    send_frame(Version091::Frames::AmqpVersion091ProtocolHeader.new)
+
+    nil
+  end
+
+  private
+
+  # Build a SASL authentication response for a username and password.
+  #
+  # @param [String] username
+  # @param [String] password
+  # @return [String]
+  def build_sasl_response_plain(username, password)
+    # per the SASL spec, username and password must be UTF-8 encoded
+    # see: https://www.rfc-editor.org/rfc/rfc4616#section-2
+    "\x00".b + username + "\x00".b + password
```

Comment:
Based on the comments above, should this ensure the strings here are already `UTF-8`?